File: java/Binary_Search/Day10/SmallestForThreshold.java
package Day10;
import java.util.*;

public class SmallestForThreshold {

    // public int smallOfDiv(int[] arr, int threshold) {
    //     int maxi = Integer.MIN_VALUE;

    //     for (int i = 0; i < arr.length; i++) {
    //         maxi = Math.max(maxi, arr[i]);
    //     }




    //     // ---------- Linear Search ----------
    //     for (int d = 1; d <= maxi; d++) {
    //         int sum = 0;

    //         for (int i = 0; i < arr.length; i++) {
    //             sum += Math.ceil((double) arr[i]/ d);
    //         }

    //         if (sum <= threshold) {
    //             return d;
    //         }
    //     }

    //     return -1;
    // }

    // public static void main(String[] args) {
    //     Scanner sc = new Scanner(System.in);

    //     System.out.print("Enter no. of arr: ");
    //     int n = sc.nextInt();

    //     int[] arr = new int[n];
    //     System.out.print("Enter the numbers: ");
    //     for (int i = 0; i < n; i++) {
    //         arr[i] = sc.nextInt();
    //     }
        

    //     System.out.print("Enter threshold: ");
    //     int threshold = sc.nextInt();

    //     SmallestForThreshold thres = new SmallestForThreshold();

    //     int ans = thres.smallOfDiv(arr, threshold);

    //     System.out.println("Smallest Divisor = " + ans);


    //     sc.close();
    // }







    public int sumByDiv(int[] arr, int div) {
        int sum = 0;
        int n = arr.length;

        for (int i = 0; i < n; i++) {
            sum += Math.ceil((double) arr[i] / (double) div);
        }

        return sum;
    }


    public int smallestDivisor(int[] arr, int threshold) {
        int n = arr.length;
        int maxi = Integer.MIN_VALUE;

        for (int i = 0; i < n; i++) {
            maxi = Math.max(maxi, arr[i]);
        }



        // ---------- Binary Search ----------
        int low = 1, high = maxi;

        while (low <= high) {
            int mid = (low + high) / 2;

            if (sumByDiv(arr, mid) <= threshold) {
                high = mid - 1;
            }
            else {
                low = mid + 1;
            }
        }

        return low;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of elements: ");
        int n = sc.nextInt();

        int[] arr = new int[n];
        System.out.print("Enter the numbers: ");
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        System.out.print("Enter threshold: ");
        int threshold = sc.nextInt();

        SmallestForThreshold obj = new SmallestForThreshold();
        int ans = obj.smallestDivisor(arr, threshold);

        System.out.println("Smallest Divisor = " + ans);


        sc.close();
    }
}

File: python/Binary_Search/day10/smallestForThreshold.py
from math import ceil

# ---------- Linear Search ----------
# def smallOfDiv(arr, threshold):
#     n = len(arr)
    # maxi = float('-inf')

    # for i in range(n):
    #     maxi = max(maxi, arr[i])

    
#     for d in range(1, maxi + 1):
#         sumD = 0

#         for i in arr:
#             sumD += ceil(i / d)

#         if sumD <= threshold:
#             return d
    
#     return -1






# ---------- Binary Search ----------
def sumByDiv(arr, div):
    n = len(arr)
    sumd = 0

    for i in arr:
        sumd += ceil(i / div)

    return sumd



def smallestDivisor(arr, threshold):
    n = len(arr)
    maxi = float('-inf')

    for i in range(n):
        maxi = max(maxi, arr[i])

    # ---------- Binary Search ----------
    low = 0
    high = maxi

    while low <= high:
        mid = (low + high) // 2

        if sumByDiv(arr, mid) <= threshold:
            high = mid - 1
        
        else:
            low = mid + 1
    
    return low




if __name__ == "__main__":
    n = int(input("Enter number of elements: "))
    
    print("Enter the numbers:", end=" ")
    arr = list(map(int, input().split()))

    threshold = int(input("Enter threshold: "))

    ans = smallestDivisor(arr, threshold)
    print("Smallest Divisor =", ans)